src/codesearch.cc

```diff
@@ -352,6 +368,49 @@ class searcher {
     friend class code_searcher::search_thread;
 };
 
+class filename_searcher {
+public:
+    filename_searcher(const code_searcher *cc,
+                      const query &q) :
+        cc_(cc), query_(&q), limiter_(q.max_matches) {}
+
+    void operator()(indexed_file *file);
+
+protected:
+    const code_searcher *cc_;
+    const query *query_;
+    thread_queue<file_result*> queue_;
+    atomic_int matches_;
+    search_limiter limiter_;
+
+    friend class code_searcher::search_thread;
+};
+
+void filename_searcher::operator()(indexed_file *file)
+{
+    if (limiter_.why())
```

Comment:
OK, I ended up refactoring the matches counter to live inside the limiter, and gave the limiter a `record_match()` method instead, which feels much nicer.

We could possibly accept the possibility of a data race and use a nonatomic int here, because:
 * this count is only used for the search limiter, so it's acceptable to be imprecise
 * limiting is already a bit imprecise--two threads can push results onto the queue in `try_match()` when strict limit enforcement would mean only one of them should
 * if correctness is a concern, we can apply limit separately when reading results from the queue later
But, there's potential for the count to much more wildly diverge from reality, and I think probably the performance win isn't significant, so I'm not going to make that change.